### ISSUE TITLE
close_popup seems not to work

### ISSUE THREAD

**Issue by nobe4** [Shougo/neocomplete.vim#412]
# Problems summary

After a fresh install of neocomplete the close_popup method seems not to work.

## Expected

The pum appears and the mappings are working (the selection with `<TAB>` for example is OK).
Bu when I try to actually do the completion, nothing appends.

## Environment Information
 * OS: OS X 10.10.3
 * Vim version: 

```
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled May 26 2015 22:56:41)
MacOS X (unix) version
Included patches: 1-383
Compiled by Homebrew
Huge version with MacVim GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_binary
+arabic          +find_in_path    -mouse_sysmouse  +tag_old_static
+autocmd         +float           +mouse_urxvt     -tag_any_white
+balloon_eval    +folding         +mouse_xterm     +tcl
+browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     +fullscreen      -mzscheme        +textobjects
+cindent         -gettext         +netbeans_intg   +title
+clientserver    -hangul_input    +odbeditor       +toolbar
+clipboard       +iconv           +path_extra      +transparency
+cmdline_compl   +insert_expand   +perl            +user_commands
+cmdline_hist    +jumplist        +persistent_undo +vertsplit
+cmdline_info    +keymap          +postscript      +virtualedit
+comments        +langmap         +printer         +visual
+conceal         +libcall         +profile         +visualextra
+cryptv          +linebreak       +python          +viminfo
+cscope          +lispindent      -python3         +vreplace
+cursorbind      +listcmds        +quickfix        +wildignore
+cursorshape     +localmap        +reltime         +wildmenu
+dialog_con_gui  +lua             +rightleft       +windows
+diff            +menu            +ruby            +writebackup
+digraphs        +mksession       +scrollbind      -X11
+dnd             +modify_fname    +signs           -xfontset
-ebcdic          +mouse           +smartindent     +xim
+emacs_tags      +mouseshape      -sniff           -xsmp
+eval            +mouse_dec       +startuptime     -xterm_clipboard
+ex_extra        -mouse_gpm       +statusline      -xterm_save
+extra_search    -mouse_jsbterm   -sun_workshop    -xpm
+farsi           +mouse_netterm   +syntax
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/Applications/MacVim.app/Contents/Resources/vim"
Compilation: clang -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_MACVIM -Wall -Wno-unknown-pragmas -pipe  -DMACOS_X_UNIX  -F/usr/local/Cellar/python/2.7.9/Frameworks -I/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/include/python2.7 -I/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/include/python2.7 -fno-strict-aliasing -fno-common -dynamic -I/usr/local/include -I/usr/local/opt/sqlite/include -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1     -I/System/Library/Frameworks/Tcl.framework/Headers  -D_REENTRANT=1  -D_THREAD_SAFE=1  -D_DARWIN_C_SOURCE=1
Linking: clang   -L. -L/usr/local/lib -L. -L/usr/local/lib -F/usr/local/Cellar/python/2.7.9/Frameworks -L/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config -lpython2.7 -framework CoreFoundation -L/usr/local/lib -o Vim -framework Cocoa -framework Carbon       -lm  -lncurses -liconv -framework Cocoa  -L/usr/local/lib -llua -fstack-protector  -L/System/Library/Perl/5.18/darwin-thread-multi-2level/CORE -lperl -framework Python  -F/System/Library/Frameworks -framework Tcl -framework CoreFoundation -framework Ruby
```

## Minimal vimrc less than 50 lines

```
set nocompatible " be iMproved, required

filetype off " required
set rtp+=~/.vim/bundle/Vundle.vim
call vundle#begin()
Bundle 'gmarik/Vundle.vim'

Bundle 'Shougo/neocomplete'
Bundle 'Shougo/neosnippet'
Bundle 'Shougo/neosnippet-snippets'

let g:neocomplete#enable_at_startup = 1 " Use neocomplete.
let g:neocomplete#enable_smart_case = 1 " Use smartcase.
let g:neocomplete#sources#syntax#min_keyword_length = 3 " Set minimum syntax keyword length.

inoremap <silent> <CR> <C-r>=<SID>my_cr_function()<CR>
function! s:my_cr_function()
    return neocomplete#close_popup() . "\<CR>"
endfunction

inoremap <expr><TAB>  pumvisible() ? "\<C-n>" : "\<TAB>"
inoremap <expr><C-h> neocomplete#smart_close_popup()."\<C-h>"
inoremap <expr><BS> neocomplete#smart_close_popup()."\<C-h>"
inoremap <expr><C-y>  neocomplete#close_popup()
inoremap <expr><C-e>  neocomplete#cancel_popup()

call vundle#end()            " required
filetype plugin indent on    " required
```

## Reproducable ways from Vim starting

 1. with the previous vimrc file saved as _vimrc launch vim : `vim -u _vimrc test.c`
 2. type `mai` then `<Tab>` to highlight the selected result 
 3. then use `<CR>`, or `<c-y>` or anything else to trigger `neocomplete#close_popup()`

## Screen shot (if possible)
[![asciicast](https://asciinema.org/a/20631.png)](https://asciinema.org/a/20631)

**Comment by Shougo** [comment]
It is feature.
If you use `neocomplete#close_popup()`, popup menu will be closed.
So your `pumvisible()` mapping does not work.

**Comment by nobe4** [comment]
OK.
What do you suggest as mapping? I used the examples from the installation guide.

**Comment by Shougo** [comment]
> What do you suggest as mapping? I used the examples from the installation guide.

Please explain your intended behavior.  I don't understand your `Expected` section.

**Comment by nobe4** [comment]
OK, sorry for the non-explicit Expected section.

For example : 
I type the start of a for loop in a javascript file (below the completion menu is symbolised by `>` ): 
```
for
> forin [nsnip] for (var ${1:i} in ${2:#:Things}) { ${0: } }
> for   [nsnip] for (var ${1:i}=0; $1 < ${2:#:Things}.length; ++$1) { ${0: }
```

Then I chose the one I want with `<Tab>`, the selected one is correctly highlighted.
Then when I press any of the `neocomplete#close_popup()` mapping the popup only close without inserting the snippet.

Meanwhile I found that I did not configure neosnippet ( I found that on #13 ).
And apparently the snippet-inserting section is handled by neosnippet.

So following the advice on [this issue](https://github.com/Shougo/neosnippet.vim/issues/254) I add the  mapping 
```
imap <expr><CR> neosnippet#expandable() ? "\<Plug>(neosnippet_expand)" : "\<CR>"
```
And it's now working perfectly.

Thanks :)

PS : if I have trouble again, how should I select the repo I post on ? (for example this one for the popup and neosnippet for snippet insertion).
